src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+	struct obj_ec_codec	*codec;
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned int		 p = entry->ae_oca->u.ec.e_p;
+	unsigned int		 cell_bytes = len * entry->ae_rsize;
+	unsigned char		*data[k];
+	unsigned char		*parity_bufs[p];
+	unsigned char		*buf;
+	int			 i;
+
+	buf = entry->ae_sgl->sg_iovs[AGG_IOV_DATA].iov_buf;
+	for (i = 0; i < k; i++)
+		data[i] = &buf[i*cell_bytes];
+
+	buf = entry->ae_sgl->sg_iovs[AGG_IOV_PARITY].iov_buf;
+	for (i = 0; i < p; i++)
+		parity_bufs[i] = &buf[i*cell_bytes];
+
+	codec = obj_ec_codec_get(daos_obj_id2class(entry->ae_oid.id_pub));
+	ec_encode_data(cell_bytes, k, p, codec->ec_gftbls, data, parity_bufs);
+}
+
+/* Driver function for full_stripe encode. Fetches the data and then invokes
+ * second function to encode the parity.
+ */
+static int
+agg_encode_local_parity(struct ec_agg_entry *entry)
+{
+	int rc = 0;
+
+	rc = agg_fetch_data_stripe(entry);
+	if (rc)
+		goto out;
+	agg_encode_full_stripe(entry);
+
+out:
+	return rc;
+}
+
+/* True if all extents within the stripe are at a higher epoch than
+ * the parity for the stripe.
+ */
+static bool
+agg_data_is_newer(struct ec_agg_entry *entry)
+{
+	struct ec_agg_extent	*agg_extent;
+
+	d_list_for_each_entry(agg_extent, &entry->ae_cur_stripe.as_dextents,
+			      ae_link) {
+		if (agg_extent->ae_epoch < entry->ae_par_extent.ape_epoch)
+			return false;
+	}
+	return true;
+}
+
+/* Determines if the replicas present for the current stripe of object entry
+ * constitute a full stripe. If parity exists for the the stripe, the replicas
+ * making up the full stripe must be a later epoch that the parity.
+ */
+static bool
+agg_stripe_is_filled(struct ec_agg_entry *entry, bool has_parity)
+{
+	bool	is_filled, rc = false;
+
+
+	is_filled = entry->ae_cur_stripe.as_stripe_fill ==
+		entry->ae_oca->u.ec.e_k * entry->ae_oca->u.ec.e_len;
+
+	if (is_filled)
+		if (!has_parity || agg_data_is_newer(entry))
+			rc = true;
+	return rc;
+}
+
+/* Writes updated parity to VOS, and removes replicas fully contained
+ * in the processed stripe.
+ */
+static int
+agg_update_vos(struct ec_agg_entry *entry)
+{
+	d_sg_list_t		sgl = { 0 };
+	daos_iod_t		iod = { 0 };
+	daos_epoch_range_t	epoch_range = { 0 };
+	d_iov_t			iov = { 0 };
+	daos_recx_t		recx = { 0 };
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	int			rc = 0;
+
+	iov.iov_buf = entry->ae_sgl->sg_iovs[AGG_IOV_PARITY].iov_buf;
+	iov.iov_buf_len =
+			entry->ae_sgl->sg_iovs[AGG_IOV_PARITY].iov_buf_len;
+	iov.iov_len =
+			entry->ae_sgl->sg_iovs[AGG_IOV_PARITY].iov_buf_len;
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr = 1;
+
+	recx.rx_idx = entry->ae_cur_stripe.as_stripenum * k * len;
+	recx.rx_nr = k * len;
+	epoch_range.epr_lo = 0ULL;
+	epoch_range.epr_hi = entry->ae_cur_stripe.as_hi_epoch;
+	rc = vos_obj_array_remove(entry->ae_chdl, entry->ae_oid,
+				  &epoch_range, &entry->ae_dkey,
+				  &entry->ae_akey, &recx);
```

Comment:
Is this for punching replicas on parity nodes? So later VOS aggregation is responsible for real deletion?

Reply:
This function performs an actual deletion in VOS.  I worked with Jeff on how best to handle these extents. At his suggestion (and due his addition of the capability to VOS), the method is used.